pkg/querier/queryrange/log_result_cache.go

```diff
@@ -0,0 +1,280 @@
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "invalid request type %T", req)
+	}
+
+	interval := validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, l.limits.QuerySplitDuration)
+	// skip caching by if interval is unset
+	if interval == 0 {
+		return l.next.Do(ctx, req)
+	}
+	// The first subquery might not be aligned.
+	alignedStart := time.Unix(0, lokiReq.GetStartTs().UnixNano()-(lokiReq.GetStartTs().UnixNano()%interval.Nanoseconds()))
+	// generate the cache key based on query, tenant and start time.
+	cacheKey := fmt.Sprintf("log:%s:%s:%d", tenant.JoinTenantIDs(tenantIDs), req.GetQuery(), alignedStart.UnixNano()/(interval.Nanoseconds()))
+
+	_, buff, _, err := l.cache.Fetch(ctx, []string{cache.HashKey(cacheKey)})
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error fetching cache", "err", err, "cacheKey", cacheKey)
+		return l.next.Do(ctx, req)
+	}
+	// we expect only one key to be found or missing.
+	if len(buff) > 1 {
+		level.Warn(l.logger).Log("msg", "unexpected length of cache return values", "buff", len(buff))
+		return l.next.Do(ctx, req)
+	}
+
+	if len(buff) == 0 {
+		// cache miss
+		return l.handleMiss(ctx, cacheKey, lokiReq)
+	}
+
+	// cache hit
+	var cachedResquest LokiRequest
+	err = proto.Unmarshal(buff[0], &cachedResquest)
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error unmarshalling request from cache", "err", err)
+		return l.next.Do(ctx, req)
+	}
+	return l.handleHit(ctx, cacheKey, &cachedResquest, lokiReq)
+}
+
+func (l *logResultCache) handleMiss(ctx context.Context, cacheKey string, req *LokiRequest) (queryrangebase.Response, error) {
+	l.metrics.CacheMiss.Inc()
+	level.Debug(l.logger).Log("msg", "cache miss", "key", cacheKey)
+	resp, err := l.next.Do(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	lokiRes, ok := resp.(*LokiResponse)
+	if !ok {
+		return nil, fmt.Errorf("unexpected response type %T", resp)
+	}
+	// At the moment we only cache empty results
+	if !isEmpty(lokiRes) {
+		return resp, nil
+	}
+	data, err := proto.Marshal(req)
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error marshalling request", "err", err)
+		return resp, nil
+	}
+	// cache the result
+	err = l.cache.Store(ctx, []string{cache.HashKey(cacheKey)}, [][]byte{data})
+	if err != nil {
+		level.Warn(l.logger).Log("msg", "error storing cache", "err", err)
+	}
+	return resp, nil
+}
+
+func (l *logResultCache) handleHit(ctx context.Context, cacheKey string, cachedResquest *LokiRequest, lokiReq *LokiRequest) (queryrangebase.Response, error) {
+	l.metrics.CacheHit.Inc()
+	// we start with an empty response
+	result := emptyResponse(cachedResquest)
+	// if the request is the same and covers the whole time range,
+	// we can just return the cached result.
+	if lokiReq.GetStartTs().After(cachedResquest.GetStartTs()) || lokiReq.GetStartTs().Equal(cachedResquest.GetStartTs()) &&
+		lokiReq.GetEndTs().Before(cachedResquest.GetEndTs()) || lokiReq.GetEndTs().Equal(cachedResquest.GetEndTs()) {
```

Comment:
nit: simplify
```suggestion
	if !lokiReq.GetStartTs().Before(cachedResquest.GetStartTs()) &&
		!lokiReq.GetEndTs().After(cachedResquest.GetEndTs()) {
```

Reply:
nit: I also wouldn't mind moving this to a method that returns a boolean and clearly states what it's doing, ie `requestWithinCachedRequestTimerange` or something.

Reply:
I would do it if it was used anywhere else, I think the comment is enough there.